### PR TITLE
Adding Payment Controllers and Payment Profile UI

### PR DESCRIPTION
**Pull request by GurkirtSingh**
### What this PR does (required):
- Creates controller to get list of payment methods for user
- Creates controller to get client secret to securely add card payment method 
- Creates controller to charge owner (requester) of pet sitting request
- Adds routes for payment controllers
- Creates Profile Payment page

### Screenshots/videos
Profile Payment Methods Settings

### Any information needed to test this feature (required):
- Login/Register
- Navigate to profile settings and payment methods
- The page should show the current saved payment methods
- click on 'Add new profile payment' will open form to enter and submit card details 
- Fetch client secret by GET `/payments/secret`
- Fetch list of cards by GET `/payments/saved-cards`
- Charge amount to requester by POST `/payments/charge-default-card` and provide request id in body of request

### Any issues with the current functionality (optional):
- The Endpoint are test through third party API platform `Postman`. Needs to integrate with front end to actual see the results

### CODE REVIEW

File: server/controllers/payment.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  const saveUser = await user.save();
+  if (!saveUser) {
+    throw new Error("Unable to link payment account to user account!!");
+  }
+  return customer;
+};
+
+// @route GET /payments/secret
+// @desc Get setup intent client secret
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/payment.js:37]: Good documentation

File: server/controllers/payment.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    res.status(404);
+    throw new Error("Unable proccess payment setup");
+  }
+  res.status(200);
+  res.send({ client_secret: setupIntent.client_secret });
+});
+
+// @route GET /payments/saved-cards
+// @desc Get list of saved cards
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/payment.js:58]: :+1:

File: server/controllers/payment.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  res.send({
+    success: {
+      cards: cards.data,
+    },
+  });
+});
+
+// @route POST /payments/charge-default-card
+// @desc collect payment from customer's saved card
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/payment.js:83]: :+1:

File: server/controllers/payment.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+exports.chargeSavedPaymentMethod = asyncHandler(async (req, res, next) => {
+  const { requestId } = req.body;
+  if (!requestId) {
```
Comment (sundayezeilo) [inline server/controllers/payment.js:86]: :+1:

File: server/controllers/payment.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  if (!cards) {
+    res.status(404);
+    throw new Error("No Card is setup yet!!");
+  }
+  res.status(200);
+  res.send({
+    success: {
+      cards: cards.data,
+    },
+  });
```
Comment (sundayezeilo) [inline server/controllers/payment.js:78]: :+1:

**Review by sundayezeilo — COMMENTED** [review]
Hey @GurkirtSingh, nice work with this one :+1:

File: client/src/helpers/APICalls/payment.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { FetchOptions } from '../../interface/FetchOptions';
+interface ClientSecretAPIData {
+  success?: {
+    client_secret?: string;
```
Comment (rajivtitus) [inline client/src/helpers/APICalls/payment.ts:4]: Please use camelCase for field names in JS
Ditto where applicable

File: server/controllers/payment.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    return await stripe.customers.retrieve(user.stripeCustomerId);
+  }
+  const customer = await stripe.customers.create({
+    email: user.email,
+    name: user.name,
+  });
+  if (!customer) {
+    throw new Error("Unable to create payment account!!");
+  }
+  user.set({ stripeCustomerId: customer.id });
```
Comment (rajivtitus) [inline server/controllers/payment.js:27]: 👍🏽

File: client/src/components/PaymentCard/PaymentCard.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  expireMonth,
+  expireYear,
+  cardHolderName,
+}) => {
+  cardHolderName = cardHolderName || '';
+  const classes = useStyles();
+  const [cardTypeLogo, setCardTypeLogo] = useState<string>();
+  useEffect(() => {
+    if (brand === 'visa') {
+      setCardTypeLogo('https://logos-world.net/wp-content/uploads/2020/04/Visa-Logo.png');
```
Comment (rajivtitus) [inline client/src/components/PaymentCard/PaymentCard.tsx:29]: Maybe consider using local images as these links may be broken in the future

File: server/controllers/payment.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+// @route POST /payments/charge-default-card
+// @desc collect payment from customer's saved card
+// @access Private
+exports.chargeSavedPaymentMethod = asyncHandler(async (req, res, next) => {
+  const { requestId } = req.body;
+  if (!requestId) {
+    res.status(400);
+    throw new Error("Please provide a request id to process payment");
+  }
+  const request = await Request.findById(requestId);
```
Comment (rajivtitus) [inline server/controllers/payment.js]:
```suggestion
  const request = await Request.findById(requestId).populate('requester');
```

**Review by rajivtitus — COMMENTED** [review]
Very good work with this Gurkirt! Please also add a picture of the payment UI to the PR template 😄 
Also, just a couple of minor comments for you to look at..

**Comment by rajivtitus** [comment]
@GurkirtSingh Please don't forget to add the payment UI screenshots/video to this PR

**Pull request merged**